### PR TITLE
Payment unit accrued and import changes

### PR DESCRIPTION
**Pull request by pxwxnvermx**
[Currency Ticket](https://dimagi-dev.atlassian.net/browse/CCPC-143)
[Payment Config Ticket](https://dimagi-dev.atlassian.net/browse/CCPC-152)

### CODE REVIEW

File: commcare_connect/opportunity/visit_import.py [hunk 1]
```diff
@@ -88,10 +90,31 @@ def _bulk_update_visit_status(opportunity: Opportunity, dataset: Dataset):
...
+        user_visits = UserVisit.objects.filter(
+            opportunity=opportunity, user=user, status=VisitValidationStatus.approved
+        ).order_by("entity_id")
+        access = OpportunityAccess.objects.get(user=user, opportunity=opportunity)
+        payment_accrued = 0
+        for payment_unit in payment_units:
+            payment_unit_deliver_units = {deliver_unit.id for deliver_unit in payment_unit.deliver_units.all()}
+            for entity_id, visits in itertools.groupby(user_visits, key=lambda x: x.entity_id):
+                deliver_units = {v.deliver_unit.id for v in visits}
+                if payment_unit_deliver_units.issubset(deliver_units) or deliver_units.issubset(
```
Comment (calellowitz) [inline commcare_connect/opportunity/visit_import.py:110]: this logic seems incorrect to me. Specifically if a payment unit contains two deliver units and the user has done one of the `deliver_units.issubset(payment_deliver_units)` will be `True` but they will not have completed the payment unit (since they only did half of it). I think you want to check only the first condition (that the payment is a subset of what is done), but I might be missing an edge case you intended.

**Review by calellowitz — CHANGES_REQUESTED** [review]
Given how complex the payment accrual logic is, this would be a great place to have some simple tests, confirming that it does what we expect given various configurations and sets of visits.

File: commcare_connect/opportunity/migrations/0024_opportunity_currency_and_more.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+class Migration(migrations.Migration):
+    dependencies = [
+        ("opportunity", "0023_paymentunit_deliverunit_payment_unit_and_more"),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="opportunity",
+            name="currency",
+            field=models.CharField(
```
[the remark this answers is not in the document]
Comment (pxwxnvermx) [reply commcare_connect/opportunity/migrations/0024_opportunity_currency_and_more.py]: Removed. https://github.com/dimagi/commcare-connect/pull/173/commits/6aca8af5f30bfc530ee71e2332ae9c9cba8d3882

File: commcare_connect/opportunity/tables.py [hunk 3]
```diff
@@ -51,16 +51,16 @@ class Meta:
...
+        fields = ("user.name", "user.username", "payment_accrued", "total_paid")
         orderable = False
         empty_text = "No payments"
 
 
 class UserStatusTable(tables.Table):
     class Meta:
         model = OpportunityAccess
-        fields = ("user.username", "accepted")
+        fields = ("user.username", "accepted", "payment_accrued", "total_paid")
```
[the remark this answers is not in the document]
Comment (pxwxnvermx) [reply commcare_connect/opportunity/tables.py]: Yeah, I was checking out the data from the UserStatus table and forgot to remove it. Fixed. https://github.com/dimagi/commcare-connect/pull/173/commits/0eda3008ee91ab799871c0f360d2f3c137dc1dce

File: commcare_connect/opportunity/views.py [hunk 4]
```diff
@@ -151,15 +150,15 @@ def get_queryset(self):
...
+    model = OpportunityAccess
     paginate_by = 25
     table_class = PaymentTable
     template_name = "tables/single_table.html"
 
     def get_queryset(self):
         opportunity_id = self.kwargs["pk"]
         opportunity = get_object_or_404(Opportunity, organization=self.request.org, id=opportunity_id)
-        return Payment.objects.filter(opportunity_access__opportunity=opportunity).order_by("date_paid")
+        return OpportunityAccess.objects.filter(opportunity=opportunity)
```
[the remark this answers is not in the document]
Comment (pxwxnvermx) [reply commcare_connect/opportunity/views.py]: Fixed. https://github.com/dimagi/commcare-connect/pull/173/commits/bdfd735a22a89d4d41d977a90f5d7128b9bdb1af

File: commcare_connect/opportunity/tests/test_visit_import.py [hunk 5]
```diff
@@ -25,6 +32,30 @@ def test_bulk_update_visit_status():
     assert after_status == {VisitValidationStatus.approved.value}
 
 
+@pytest.mark.django_db
```
Comment (calellowitz) [inline commcare_connect/opportunity/tests/test_visit_import.py:31]: nit: fun fact that simon taught me is that you don't need these annotations if your test depends on fixtures that depend on the `db` fixture, as these two fixtures do

File: commcare_connect/opportunity/tests/test_visit_import.py [hunk 6]
```diff
@@ -25,6 +28,30 @@ def test_bulk_update_visit_status():
...
+                user=mobile_user,
+                deliver_unit=deliver_unit,
+                entity_id=deliver_unit.payment_unit.name,
+                status=VisitValidationStatus.approved.value,
+            )
+        )
+    update_payment_accrued(opportunity, {mobile_user.id})
+    assert OpportunityAccess.objects.filter(user=mobile_user, opportunity=opportunity).exists()
+    access = OpportunityAccess.objects.get(user=mobile_user, opportunity=opportunity)
+    assert access.payment_accrued == sum(payment_unit.amount for payment_unit in payment_units)
```
Comment (calellowitz) [inline commcare_connect/opportunity/tests/test_visit_import.py:52]: Nit: it would be nice to have a test that confirmed the user can get a payment unit multiple times (met the criteria for different entities)

**Review by calellowitz — APPROVED** [review]
